processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionExecutor.java

```diff
@@ -140,25 +139,72 @@ public CarbonCompactionExecutor(Map<String, TaskBlockInfo> segmentMapping,
+        // , then make a single RawResultIterator for all the blocks
+        List<List<TableBlockInfo>> listOfTableBlocksBasedOnKeyLength =
+            getListOfTableBlocksBasedOnKeyLength(list);
+        for (List<TableBlockInfo> tableBlockInfoList : listOfTableBlocksBasedOnKeyLength) {
+          Collections.sort(tableBlockInfoList);
+          LOGGER.info(
+              "for task -" + task + "- in segment id -" + segmentId + "- block size is -" + list
+                  .size());
+          queryModel.setTableBlockInfos(tableBlockInfoList);
+          if (sortingRequired) {
+            resultList.get(CarbonCompactionUtil.UNSORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          } else {
+            resultList.get(CarbonCompactionUtil.SORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          }
         }
       }
     }
     return resultList;
   }
 
+  /**
+   * This method returns the List of TableBlockInfoList, where each listOfTableBlockInfos will have
+   * same keySize
+   * @param tableBlockInfos List of tableBlock Infos
+   * @return
+   */
+  private List<List<TableBlockInfo>> getListOfTableBlocksBasedOnKeyLength(
+      List<TableBlockInfo> tableBlockInfos) {
+    List<List<TableBlockInfo>> listOfTableBlockInfoListOnKeySize = new ArrayList<>();
+    // create a map of keySizeInBytes got from dataFileFooter
+    Map<Integer, List<TableBlockInfo>> keySizeToTableBlockInfoMap = new HashMap<>();
+    // iterate through tableBlockInfos
+    for (TableBlockInfo tableBlock : tableBlockInfos) {
+      // get the keySizeInBytes for the dataFileFooter
+      int keySizeInBytes =
+          getSourceSegmentProperties(Collections.singletonList(tableBlock.getDataFileFooter()))
+              .getDimensionKeyGenerator().getKeySizeInBytes();
+      // get the value for the key keySizeInBytes
+      List<TableBlockInfo> tempBlockInfoList = keySizeToTableBlockInfoMap.get(keySizeInBytes);
+      if (tempBlockInfoList == null) {
+        // if the list is null we haven't seen any list of TableBlockInfo with this KeySize before,
+        // create a new List and add to map
+        tempBlockInfoList = new ArrayList<>();
+        keySizeToTableBlockInfoMap.put(keySizeInBytes, tempBlockInfoList);
+      }
+      // add the tableBlockInfo to list
+      tempBlockInfoList.add(tableBlock);
+    }
+    for (Map.Entry<Integer, List<TableBlockInfo>> taskMap : keySizeToTableBlockInfoMap.entrySet()) {
```

Comment:
No need to change into list<list>, return map<list> directly.  And the line can print the  keysize

Reply:
printing key size is not much useful info may be, this can be internal handling, it is ok i think